crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -193,6 +193,20 @@ pub(crate) fn infer_expression_types<'db>(
     TypeInferenceBuilder::new(db, InferenceRegion::Expression(expression), index).finish()
 }
 
+// Similar to `infer_expression_types` (with the same restrictions). Directly returns the
+// type of the overall expression. This is a salsa query because it accesses `node_ref`,
+// which is sensitive to changes in the AST. Making it a query allows downstream queries
+// to short-circuit if the result type has not changed.
```

Comment:
We may want to add a note that it isn't necessary to use this query in `TypeInferenceBuilder` or anywhere else where it's known that the `expression` belongs to the current file.